### DISCUSSION TITLE
Add an option to middleware::Logger to log requests as they're received

### DISCUSSION THREAD

**Discussion by Castavo** [actix/actix-web#3125] [Ideas]
It would be really nice to have an option to log requests as they're received by my app:
* it would tell me what the server is doing in case of time-consuming requests
* in case the server crashed, it would tell me what the last request being processed were

**Comment by robjtede** [comment]
You can very easily write a middleware that does this.

Obviously this is going to be somewhat more limited than the built in logger because it can't show HTTP version, response status codes or sizes on the same line.

```rust
async fn pre_logger(
    req: ServiceRequest,
    next: Next<impl MessageBody + 'static>,
) -> Result<ServiceResponse<impl MessageBody>, Error> {
    tracing::info!(
        "{} \"{} {}\" \"{}\" \"{}\"",
        req.peer_addr().unwrap(),
        req.method(),
        req.path(),
        req.headers().get("referer").and_then(|hdr| hdr.to_str().ok()).unwrap_or("-"),
        req.headers().get("user-agent").and_then(|hdr| hdr.to_str().ok()).unwrap_or("-"),
    );

    next.call(req).await
}

App::new()
  .wrap(from_fn(pre_logger))
```


(Uses [`from_fn`](https://docs.rs/actix-web-lab/latest/actix_web_lab/middleware/fn.from_fn.html))
